src/main/scala/viper/gobra/translator/library/outlines/OutlinesImpl.scala

```diff
@@ -0,0 +1,135 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) 2011-2020 ETH Zurich.
+
+package viper.gobra.translator.library.outlines
+
+import viper.gobra.translator.util.ViperUtil
+import viper.silver.ast.Member
+import viper.silver.{ast => vpr}
+
+class OutlinesImpl extends Outlines {
+
+  override def finalize(addMemberFn: Member => Unit): Unit = {
+    generatedMembers foreach addMemberFn
+  }
+  private var generatedMembers: List[vpr.Member] = List.empty
+
+  /**
+    * Moves code `body` into a generated method with name `name`, preconditions `pres`, and postconditions `pros`.
+    * Returns a call to the generated method with the appropriate arguments and results.
+    *
+    * outline
+    *  requires PRE
+    *  ensures  POST
+    * (
+    *   BODY
+    * )
+    *
+    * is transformed to
+    *
+    * results := name(arguments)
+    *
+    * with
+    *
+    * method name(arguments') returns (results')
+    *   requires PRE[arguments -> arguments']
+    *   ensures  POST[results -> results'][arguments -> arguments'][ old[l] -> old ]
+    * {
+    *    var arguments
+    *    arguments := arguments'
+    *    BODY[ old[l] -> old ]
+    *    results' := results
+    * }
+    *
+    * where
+    *   arguments := free(BODY) + free(PRE) + free(Post)  // variables used in body and spec
```

Comment:
`free(Post)` here also includes the variables that are declared in the body right? In that case, the choice of values passed in the place of the corresponding parameters is inconsequential?

Reply:
free(POST) are only the free vaiables in the postcondition. I should probably substract the modified variables. Note that a parameter may only be read in the postcondition.

Reply:
I just noticed that I do not have to substract the modified variables because they are a subset of the free body variables.